### PR TITLE
Connection data for context is automatically freed when context is destroyed

### PR DESCRIPTION
**Pull request by healytpk**
In my network analysis program, **_Dynamo_**, I have a thread dedicated to sending packets out on the network. This thread runs through a few coroutines in series, and one of the coroutines sends a '**_Samba Enum List_**' query to every host on the network. The asynchronous API in **_libsmb2_** is perfect for this.
However I have coded a timeout into **_Dynamo_**. If the list of shares hasn't been received within 5 seconds, the connection to the Samba server on TCP port 445 is instantly closed by calling '**_smb2_destroy_context_**'. This can cause a memory leak of the '**_connection_data_**' structure that gets allocated inside '**_smb2_connect_share_async_**'.
The purpose of this patch is to allow instantaneous closure of the TCP socket without a memory leak.

I have minimal experience with **_libsmb2_**, I only started using it 2 days ago -- so maybe I'm just using the library incorrectly and causing the memory leak myself.

### CODE REVIEW

File: lib/init.c [hunk 1]
```diff
@@ -362,6 +362,13 @@ void smb2_destroy_context(struct smb2_context *smb2)
         free(discard_const(smb2->workstation));
         free(smb2->enc);
 
+        if ( NULL != smb2->connect_data )
```
Comment (sahlberg) [inline lib/init.c:365]: Please match the style of the rest of the file.  The curly branch goes on the same line as the if-statement
, no spaces after (,  no space before ) and we don't compare != NULL. Instead this line should just be

if (smb2->connect_data) {

File: lib/libsmb2.c [hunk 2]
```diff
@@ -526,6 +526,8 @@ free_c_data(struct smb2_context *smb2, struct connect_data *c_data)
         free(discard_const(c_data->share));
         free(discard_const(c_data->user));
         free(c_data);
+
+        if ( smb2->connect_data == c_data ) smb2->connect_data = NULL;  /* to prevent double-free in smb2_destroy_context */
```
Comment (sahlberg) [inline lib/libsmb2.c:530]: We always use curly braces even if the body of the if-statement is just a single statement/expression, so this should be

```
if (smb2->connect_data == c_data) {
    smb2->connect_data = NULL;
}
```

**Review by sahlberg — COMMENTED** [review]
You found a real memory leak. Just make these small style changes I suggested and it should be good.

File: lib/init.c (hunk 1 above)
Comment (healytpk) [reply lib/init.c]: done

File: lib/init.c [hunk 3]
```diff
@@ -362,6 +362,13 @@ void smb2_destroy_context(struct smb2_context *smb2)
...
+        {
+            struct connect_data;                                                  /* defined in libsmb2.c            */
+            extern void free_c_data(struct smb2_context*, struct connect_data*);  /* defined in libsmb2.c            */
```
[the remark this answers is not in the document]
Comment (healytpk) [reply lib/init.c]: done

**Comment by sahlberg** [comment]
Since the first commit is not yet merged, lets combine these into a single commit instead.

In git you do that with "git rebase -i HEAD~2"
and then you change the second commit from "pick" to "f"  and save.
That will merge the two commits into a single commit instead.

The re-push to github    using "git push -f"   (-f since your local repo and githubs repo have diverged)

**Comment by sahlberg** [comment]
I see you are struggling with updating the pull request. Don't worry about it. I can just cherry-pick the patch that you created tomorrow.

**Pull request closed without merging**